include/kamping/p2p/sendrecv.hpp

```diff
@@ -0,0 +1,267 @@
+
+    KASSERT(
+        !(send_buf.size() == recv_buf.size() && send_buf.data() == recv_buf.data()),
+        "Send buffer and recv buffer can not be the same."
+    );
+
+    using default_recv_count_type = decltype(kamping::recv_count_out());
+    auto recv_count_param =
+        internal::select_parameter_type_or_default<internal::ParameterType::recv_count, default_recv_count_type>(
+            std::tuple(),
+            args...
+        )
+            .construct_buffer_or_rebind();
+
+    constexpr bool is_recv_serialization_used = internal::buffer_uses_serialization<decltype(recv_buf)>;
+
+    using recv_value_type = typename std::remove_reference_t<decltype(recv_buf)>::value_type;
+    static_assert(
+        !std::is_same_v<recv_value_type, internal::unused_tparam>,
+        "No recv_buf parameter provided and no receive value given as template parameter. One of these is required."
+    );
+    auto recv_type = internal::determine_mpi_recv_datatype<recv_value_type, decltype(recv_buf)>(args...);
+    [[maybe_unused]] constexpr bool recv_type_is_in_param = !internal::has_to_be_computed<decltype(recv_type)>;
+
+    using default_recv_tag_type = decltype(kamping::tag(tags::any));
+    auto&& recv_tag_param =
+        internal::select_parameter_type_or_default<internal::ParameterType::recv_tag, default_recv_tag_type>(
+            {},
+            args...
+        );
+    constexpr auto tag_type = std::remove_reference_t<decltype(recv_tag_param)>::tag_type;
+    if constexpr (tag_type == internal::TagType::value) {
+        int recv_tag = recv_tag_param.tag();
+        KASSERT(
+            Environment<>::is_valid_tag(recv_tag),
+            "invalid recv tag " << recv_tag << ", must be in range [0, " << Environment<>::tag_upper_bound() << "]"
+        );
+    }
+
+    using default_status_param_type = decltype(kamping::status(kamping::ignore<>));
+    auto status =
+        internal::select_parameter_type_or_default<internal::ParameterType::status, default_status_param_type>(
+            {},
+            args...
+        )
+            .construct_buffer_or_rebind();
+
+    using default_source_buf_type = decltype(kamping::source(rank::any));
+    auto&& source_param =
+        internal::select_parameter_type_or_default<internal::ParameterType::source, default_source_buf_type>(
+            {},
+            args...
+        );
+
+    KASSERT(internal::is_valid_rank_in_comm(source_param, *this, true, true));
+    int source = source_param.rank_signed();
+
+    // Calculate rev_count if not given by calling sendrecv with the send_count
+    if constexpr (internal::has_to_be_computed<decltype(recv_count_param)>) {
+        this->sendrecv(
+            kamping::destination(destination.rank_signed()),
+            kamping::send_count(1),
+            kamping::send_buf(send_count.get_single_element()),
+            kamping::recv_count(1),
+            kamping::recv_buf(recv_count_param)
```

Comment:
Can you really receive into the parameter object. I am not really sure if that works, so better pass an lvalue-ref to an int here, and the assign it to the params internal value.